File: frontend/src/game/upgrades/stat_modifiers/tenacity_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Tenacity1Def: UpgradeDef = {
  id: "tenacity_1",
  name: "Tenacity",
  description: "Increased the bullet lifespan by .15 seconds",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.TimeLeft,
  value: 150,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 6,
}

export class Tenacity1 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/tenacity_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Tenacity1(UpgradeImplementation):
    id = 'tenacity_1'
    name = 'Tenacity'
    description = 'Increased the bullet lifespan by .15 seconds'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 150
        appliers.modifiers.add_multiplier('bullet', 'timeLeft', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
